substrate/primitives/runtime/src/generic/checked_extrinsic.rs

```diff
@@ -18,81 +18,115 @@
 //! Generic implementation of an extrinsic that has passed the verification
 //! stage.
 
+use codec::Encode;
+
 use crate::{
 	traits::{
-		self, DispatchInfoOf, Dispatchable, MaybeDisplay, Member, PostDispatchInfoOf,
-		SignedExtension, ValidateUnsigned,
+		self, transaction_extension::TransactionExtension, DispatchInfoOf, DispatchTransaction,
+		Dispatchable, MaybeDisplay, Member, PostDispatchInfoOf, ValidateUnsigned,
 	},
 	transaction_validity::{TransactionSource, TransactionValidity},
 };
 
+/// The kind of extrinsic this is, including any fields required of that kind. This is basically
+/// the full extrinsic except the `Call`.
+#[derive(PartialEq, Eq, Clone, sp_core::RuntimeDebug)]
+pub enum ExtrinsicFormat<AccountId, Extension> {
+	/// Extrinsic is bare; it must pass either the bare forms of `TransactionExtension` or
+	/// `ValidateUnsigned`, both deprecated, or alternatively a `ProvideInherent`.
+	Bare,
+	/// Extrinsic has a default `Origin` of `Signed(AccountId)` and must pass all
+	/// `TransactionExtension`s regular checks and includes all extension data.
+	Signed(AccountId, Extension),
+	/// Extrinsic has a default `Origin` of `None` and must pass all `TransactionExtension`s.
+	/// regular checks and includes all extension data.
+	General(Extension),
+}
+
+// TODO: Rename ValidateUnsigned to ValidateInherent
```

Comment:
I see that you have made follow-up issues, but don't see these being addressed in any of them. Would be good to have an issue for them too, even if they are small. They can be grouped into one issue.

Reply:
This is a bit tough because the next step in [Extrinsic Horizon](#2415) is to get rid of unsigned transactions, and `ValidateUnsigned` would have to be removed and its checks moved somewhere else. With step 1 done and before step 2, renaming `ValidateUnsigned` to `ValidateInherent` wouldn't really make sense because unsigned transactions still exist and would be validated using this trait. This will probably just be removed in step 2, but we should clear up these todo's, which basically just describe a subset of the changes required in step 2.